### ISSUE TITLE
[native-image] Incompatible change of initialization policy: trying to change RERUN

### ISSUE THREAD

**Issue by aclement** [oracle/graal#1462]
MacbookPro: Graal 19.1.0

I'm upgrading my spring boot feature with Graal 19.1 support. But something has changed and I don't understand the latest message mentioning RERUN.

I'm just doing a very basic app which works with my feature and 19.0.x:

With existing feature (with minor changes to upgrade to latest APIs which have moved package): 

```
Error: Class initialization of sun.reflect.misc.Trampoline failed. Use the option 
--initialize-at-run-time=sun.reflect.misc.Trampoline to explicitly request delayed 
initialization of this class.
```

With the existing feature, Trampoline was explicitly in the buildTimeInitialization list.
So I removed Trampoline from the explicit list.

```
Error: No instances are allowed in the image heap for a class that is initialized or 
reinitialized at image runtime: com.sun.jmx.mbeanserver.MXBeanLookup. Try marking this class for build-time
initialization with --initialize-at-build-time=com.sun.jmx.mbeanserver.MXBeanLookup
```

Added MXBeanLookup to build time initialization list:

```
  Error: Incompatible change of initialization policy for com.sun.jmx.mbeanserver.MXBeanLookup: 
trying to change RERUN ensure com.sun.jmx.mbeanserver.MXBeanLookup#mbscToLookup
clean state to BUILD_TIME from feature org.springframework.boot.graal.support.InitializationHandler.register 
```

Any advice?

**Comment by vjovanov** [comment]
Could you send me the full reproducer? That would make it easier for me to fix the issue.

**Comment by aclement** [comment]
Hey, thanks for taking a look. Ok, this should do it (unless you are on windows...):
```
git clone git@github.com:aclement/spring-boot-graal-feature.git 
cd spring-boot-graal-feature
git checkout graal_19_1_0
mvn clean package
cd samples/commandlinerunner
./compile.sh
```
Some of the config is statically defined and passed in to the image builder, other stuff is computed (by chasing down spring configuration). If you want to tweak the static stuff, see the files in `src/main/resources` (and rebuild the feature afterwards).

It could well be something I'm doing wrong but I don't quite understand the message to know what to change.

**Comment by vjovanov** [comment]
The problem here is not in class initialization mechanism itself. It actually exactly says what is the problem: Our JDK support marks `com.sun.jmx.mbeanserver.MXBeanLookup` as `RERUN` and you try to make it `BUILD_TIME`.

I think the problem here is that `RERUN` should not be used for `com.sun.jmx.mbeanserver.MXBeanLookup`. I think we should recompute field values here if we want to clean state and we should initialize this class at build-time.

Do you have an idea of how `com.sun.jmx.mbeanserver.MXBeanLookup` ends up in the image heap?

**Comment by aclement** [comment]
If I remove it (from build time init) I get this:
```

Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: No instances are allowed in the image heap for a class that is initialized or reinitialized at image runtime: com.sun.jmx.mbeanserver.MXBeanLookup. Try marking this class for build-time initialization with --initialize-at-build-time=com.sun.jmx.mbeanserver.MXBeanLookup
Detailed message:
Trace:  object com.sun.jmx.mbeanserver.MXBeanSupport
        object javax.management.StandardMBean
        object com.sun.jmx.mbeanserver.NamedObject
        object java.util.HashMap$Node
        object java.util.HashMap$Node[]
        object java.util.HashMap
        object java.util.HashMap$Node
        object java.util.HashMap$Node
        object java.util.HashMap$Node[]
        object java.util.HashMap
        object com.sun.jmx.mbeanserver.Repository
        object com.sun.jmx.interceptor.DefaultMBeanServerInterceptor
        object com.sun.jmx.mbeanserver.JmxMBeanServer
        object java.lang.Object[]
        object java.util.ArrayList
        method javax.management.MBeanServerFactory.removeMBeanServer(MBeanServer)
Call path from entry point to javax.management.MBeanServerFactory.removeMBeanServer(MBeanServer): 
        at javax.management.MBeanServerFactory.removeMBeanServer(MBeanServerFactory.java:422)
        at javax.management.MBeanServerFactory.releaseMBeanServer(MBeanServerFactory.java:154)
        at org.springframework.jmx.support.MBeanServerFactoryBean.destroy(MBeanServerFactoryBean.java:210)
        at org.springframework.beans.factory.support.DisposableBeanAdapter.destroy(DisposableBeanAdapter.java:256)
        at org.springframework.beans.factory.support.DisposableBeanAdapter.run(DisposableBeanAdapter.java:233)
        at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:144)
        at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:89)
        at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:139)
        at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

        at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:130)
        at com.oracle.graal.pointsto.BigBang.finish(BigBang.java:565)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:683)
```

which I guess indicates it comes in due to `org.springframework.jmx.support.MBeanServerFactoryBean.destroy(MBeanServerFactoryBean.java:210)`

**Comment by vjovanov** [comment]
@olpaw it seems that we can't use rerun for `com.sun.jmx.mbeanserver.MXBeanLookup`

**Comment by aclement** [comment]
I used some substitutes to temporarily get around the code bringing it in (as I don't actually need it for this app). Nice work on the compile times!
```
19.0.2:  [total]:  98,917.96 ms
19.1.0:  [total]:  75,521.97 ms
```

**Comment by olpaw** [comment]
Fixed on master: 5b48670281438f93739f73b62d3ddb250d7eb974

**Comment by olpaw** [comment]
Also created backport ticket to get the change into `19.1.1`
cc @ezzarghili
